README.md
# image-layout
[![NPM version](http://img.shields.io/npm/v/image-layout.svg?style=flat)](https://www.npmjs.org/package/image-layout)

A collection of deterministic image layout algorithms written in pure javascript that work just as well on a server as in a browser. None of them are tied to the DOM in any way.

```sh
$ npm install image-layout --save
```

### Sample Usage
```js
var layout = require('image-layout/layouts/fixed-partition');
var elements = [
	{width: 400, height: 300},
	{width: 200, height: 100}
];

var result = layout(elements);
// container_width = result.width
// container_height = result.height
for (var i = 0, n = elements.length; i < n; i++) {
	// result.positions[i].x
	// result.positions[i].y
	// result.positions[i].width
	// result.positions[i].height
}
```

### Layouts

#### fixed-columns

Simply balances photos into multiple columns.

```js
var layout = require('image-layout/layouts/fixed-columns');
var result = layout(elements, {
    containerWidth: 800,
    columnCount: 3,
    spacing: 0
});
```

<img src="./docs/fixed-columns.jpg?raw=true" width="500" />

#### fixed-partition

Uses a partitioning scheme outlined by [Johannes Treitz](https://twitter.com/jtreitz) in ["The algorithm for a perfectly balanced photo gallery"](https://www.crispymtn.com/stories/the-algorithm-for-a-perfectly-balanced-photo-gallery). It's not the most-efficient option, but it leaves no ragged edges.

```js
var layout = require('image-layout/layouts/fixed-partition');
var result = layout(elements, {
    align: 'center', // optional
    containerWidth: 800,
    idealElementHeight: 300,
    spacing: 0
});
```

<img src="./docs/fixed-partition.jpg?raw=true" width="500" />


#### single

Constrains a single image to a given area.

```js
var layout = require('image-layout/layouts/single');
var result = layout(elements, {
    align: 'center', // optional
    containerWidth: 800,
    maxHeight: 800
});
```

## License

Copyright &copy; 2015 [Natural Atlas, Inc.](https://github.com/naturalatlas) & [Contributors](https://github.com/naturalatlas/image-layout/graphs/contributors)

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at: http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
See https://github.com/naturalatlas/image-layout/releases

layouts/single.js
/**
 * Algorithm: fixed-columns
 *
 * Options:
 *   - containerWidth Max width of the parent container (in pixels)
 *   - maxHeight Max height of the parent container (in pixels)
 *
 * @throws
 * @param {object[]} elements
 * @param {object} options
 * @return {object}
 */
module.exports = function(elements, options) {
	var el = elements[0];

	var scale = 1;
	if (el.width * scale > options.maxWidth) scale = options.maxWidth / el.width;
	if (el.height * scale > options.containerWidth) scale *= options.containerWidth / (scale * el.height);

	var w = Math.round(el.width * scale);
	var h = Math.round(el.height * scale);
	var padLeft = options.align === 'center' ? Math.floor((options.containerWidth - w)/2) : 0;

	return {
		width: options.containerWidth,
		height: h,
		positions: [{x: padLeft, y: 0, width: w, height: h}]
	};
};


layouts/fixed-columns.js
/**
 * Algorithm: fixed-columns
 *
 * Options:
 *   - containerWidth   Width of the parent container (in pixels)
 *   - columnCount      Number of columns
 *   - spacing          Spacing between items (in pixels)
 *
 * @throws
 * @param {object[]} elements
 * @param {object} options
 * @return {object}
 */
module.exports = function(elements, options) {
	var i, j, l, n, k, limit, i_column,col, colMin, src_width, src_height, dst_width, dst_height, aspect;

	var spacing = options.spacing || 0;
	var containerWidth = options.containerWidth;
	if (!containerWidth) throw new Error('Invalid container width');
	var columnCount = options.columnCount || 3;
	var columnWidth = Math.round((containerWidth - (columnCount - 1) * spacing) / columnCount);
	var maxSpan = options.maxSpan || 3;

	var positions = [];
	var cols = [];
	for (i = 0; i < columnCount; i++) {
		cols.push(0);
	}

	// distribute images to columns as evenly as possible
	for (i = 0, n = elements.length; i < n; i++) {
		src_width = elements[i].width;
		src_height = elements[i].height;
		aspect = src_width / src_height;
		dst_width = columnWidth;
		dst_height = Math.round(dst_width / aspect);

		// pick the column that is least-filled
		colMin = false;
		for (i_column = 0, k = columnCount; i_column < k; i_column++) {
			if (colMin === false || cols[i_column] < colMin) {
				colMin = cols[i_column];
				col = i_column;
			}
		}

		// update the column heights
		var y = cols[col];
		var adj = (y !== 0) ? spacing : 0;
		y += adj;
		cols[col] += dst_height + adj;

		positions.push({
			x: (col * columnWidth + col * spacing),
			y: y,
			width: dst_width,
			height: dst_height
		});
	}

	// TODO: equalize columns w/option?
	return {
		width: containerWidth,
		height: Math.max.apply(Math, cols),
		positions: positions
	}
};


layouts/fixed-partition.js
/**
 * Algorithm: fixed-partition
 *
 * The algorithm outlined by Johannes Treitz in "The algorithm
 * for a perfectly balanced photo gallery" (see url below).
 *
 * Options:
 *   - containerWidth      Width of the parent container (in pixels)
 *   - idealElementHeight  Ideal element height (in pixels)
 *   - spacing             Spacing between items (in pixels)
 *
 * @throws
 * @see https://www.crispymtn.com/stories/the-algorithm-for-a-perfectly-balanced-photo-gallery
 * @param {object[]} elements
 * @param {object} options
 * @return {object}
 */
module.exports = function(elements, options) {
	var i, n, positions = [], elementCount;

	var spacing = options.spacing || 0;
	var containerWidth = options.containerWidth;
	var idealHeight = options.idealElementHeight || (containerWidth / 3);
	if (!containerWidth) throw new Error('Invalid container width');

	// calculate aspect ratio of all photos
	var aspect;
	var aspects = [];
	var aspects100 = [];
	for (i = 0, n = elements.length; i < n; i++) {
		aspect = elements[i].width / elements[i].height;
		aspects.push(aspect);
		aspects100.push(Math.round(aspect * 100));
	}

	// calculate total width of all photos
	var summedWidth = 0;
	for (i = 0, n = aspects.length; i < n; i++) {
		summedWidth += aspects[i] * idealHeight;
	}

	// calculate rows needed
	var rowsNeeded = Math.round(summedWidth / containerWidth)

	// adjust photo sizes
	if (rowsNeeded < 1) {
		// (2a) Fallback to just standard size
		var xSum = 0, width;
		elementCount = elements.length;

		var padLeft = 0;
		if (options.align === 'center') {
			var spaceNeeded = (elementCount-1)*spacing;
			for (var i = 0; i < elementCount; i++) {
				spaceNeeded += Math.round(idealHeight * aspects[i]) - (spacing * (elementCount - 1) / elementCount);
			}
			padLeft = Math.floor((containerWidth - spaceNeeded) / 2);
		}

		for (var i = 0; i < elementCount; i++) {
			width = Math.round(idealHeight * aspects[i]) - (spacing * (elementCount - 1) / elementCount);
			positions.push({
				y: 0,
				x: padLeft + xSum,
				width: width,
				height: idealHeight
			});
			xSum += width;
			if (i !== n - 1) {
				xSum += spacing;
			}
		}
		ySum = idealHeight;
	} else {
		// (2b) Distribute photos over rows using the aspect ratio as weight
		var partitions = linear_partition(aspects100, rowsNeeded);
		var index = 0;
		var ySum = 0, xSum;
		for (i = 0, n = partitions.length; i < n; i++) {
			var element_index = index;
			var summedRatios = 0;
			for (j = 0, k = partitions[i].length; j < k; j++) {
				summedRatios += aspects[element_index + j];
				index++;
			}

			xSum = 0;
			height = Math.round(containerWidth / summedRatios);
			elementCount = partitions[i].length;
			for (j = 0; j < elementCount; j++) {
				width = Math.round((containerWidth - (elementCount - 1) * spacing) / summedRatios * aspects[element_index + j]);
				positions.push({
					y: ySum,
					x: xSum,
					width: width,
					height: height
				});
				xSum += width;
				if (j !== elementCount - 1) {
					xSum += spacing;
				}
			}
			ySum += height;
			if (i !== n - 1) {
				ySum += spacing;
			}
		}
	}

	return {
		width: containerWidth,
		height: ySum,
		positions: positions
	};
};

/**
 * Partitions elements into rows.
 *
 * @author Johannes Treitz <https://twitter.com/jtreitz>
 * @see https://www.crispymtn.com/stories/the-algorithm-for-a-perfectly-balanced-photo-gallery
 * @param {int[]} seq
 * @param {int} k
 * @return {int[][]}
 */
var linear_partition = function(seq, k) {
	var ans, i, j, m, n, solution, table, x, y, _i, _j, _k, _l;
	var _m, _nn;

	n = seq.length;
	if (k <= 0) {
		return [];
	}
	if (k > n) {
		return seq.map(function(x) {
			return [x];
		});
	}
	table = (function() {
		var _i, _results;
		_results = [];
		for (y = _i = 0; 0 <= n ? _i < n : _i > n; y = 0 <= n ? ++_i : --_i) {
			_results.push((function() {
				var _j, _results1;
				_results1 = [];
				for (x = _j = 0; 0 <= k ? _j < k : _j > k; x = 0 <= k ? ++_j : --_j) {
					_results1.push(0);
				}
				return _results1;
			})());
		}
		return _results;
	})();
	solution = (function() {
		var _i, _ref, _results;
		_results = [];
		for (y = _i = 0, _ref = n - 1; 0 <= _ref ? _i < _ref : _i > _ref; y = 0 <= _ref ? ++_i : --_i) {
			_results.push((function() {
				var _j, _ref1, _results1;
				_results1 = [];
				for (x = _j = 0, _ref1 = k - 1; 0 <= _ref1 ? _j < _ref1 : _j > _ref1; x = 0 <= _ref1 ? ++_j : --_j) {
					_results1.push(0);
				}
				return _results1;
			})());
		}
		return _results;
	})();
	for (i = _i = 0; 0 <= n ? _i < n : _i > n; i = 0 <= n ? ++_i : --_i) {
		table[i][0] = seq[i] + (i ? table[i - 1][0] : 0);
	}
	for (j = _j = 0; 0 <= k ? _j < k : _j > k; j = 0 <= k ? ++_j : --_j) {
		table[0][j] = seq[0];
	}
	for (i = _k = 1; 1 <= n ? _k < n : _k > n; i = 1 <= n ? ++_k : --_k) {
		for (j = _l = 1; 1 <= k ? _l < k : _l > k; j = 1 <= k ? ++_l : --_l) {

			m = [];
			for (x = _m = 0; 0 <= i ? _m < i : _m > i; x = 0 <= i ? ++_m : --_m) {
				m.push([Math.max(table[x][j - 1], table[i][0] - table[x][0]), x]);
			}

			var minValue, minIndex = false;
			for (_m = 0, _nn = m.length; _m < _nn; _m++) {
				if (_m === 0 || m[_m][0] < minValue) {
					minValue = m[_m][0];
					minIndex = _m;
				}
			}

			m = m[minIndex];
			table[i][j] = m[0];
			solution[i - 1][j - 1] = m[1];
		}
	}
	n = n - 1;
	k = k - 2;
	ans = [];
	while (k >= 0) {
		ans = [
			(function() {
				var _m, _ref, _ref1, _results;
				_results = [];
				for (i = _m = _ref = solution[n - 1][k] + 1, _ref1 = n + 1; _ref <= _ref1 ? _m < _ref1 : _m > _ref1; i = _ref <= _ref1 ? ++_m : --_m) {
					_results.push(seq[i]);
				}
				return _results;
			})()
		].concat(ans);
		n = solution[n - 1][k];
		k = k - 1;
	}
	return [
		(function() {
			var _m, _ref, _results;
			_results = [];
			for (i = _m = 0, _ref = n + 1; 0 <= _ref ? _m < _ref : _m > _ref; i = 0 <= _ref ? ++_m : --_m) {
				_results.push(seq[i]);
			}
			return _results;
		})()
	].concat(ans);
};


examples/index.js
function generateGrid(layoutName, options, images) {
	var layout = require('../layouts/' + layoutName + '.js');

	// calculate positioning
	var photoset = document.createElement('div');
	photoset.className = 'photos';
	var result = layout(images, options);

	// build html
	var elements = [];
	var positions = result.positions;
	for (var i = 0, n = positions.length; i < n; i++) {
		elements.push('<div style="background-image:url(' + images[i].src + ');width:' + positions[i].width + 'px;height:' + positions[i].height + 'px;left:' + positions[i].x + 'px;top:' + positions[i].y + 'px;position:absolute;"></div>');
	}
	photoset.innerHTML = elements.join('');
	photoset.style.width = result.width + 'px';
	photoset.style.height = result.height + 'px';

	var container = document.createElement('div');
	container.innerHTML += '<h2><code>' + layoutName + '</code></h2>';
	container.appendChild(photoset);
	document.body.appendChild(container);
};

generateGrid('fixed-columns', {
	containerWidth: 500,
	columnCount: 3,
	spacing: 5
}, [
	{src: 'photos/1.jpg', width: 600/2, height: 400/2},
	{src: 'photos/2.jpg', width: 400/2, height: 600/2},
	{src: 'photos/3.jpg', width: 480/2, height: 600/2},
	{src: 'photos/4.jpg', width: 600/2, height: 355/2},
	{src: 'photos/5.jpg', width: 600/2, height: 400/2},
	{src: 'photos/6.jpg', width: 600/2, height: 400/2},
	{src: 'photos/7.jpg', width: 600/2, height: 366/2}
]);

generateGrid('fixed-partition', {
	containerWidth: 500,
	idealElementHeight: 150,
	spacing: 5
}, [
	{src: 'photos/1.jpg', width: 600/2, height: 400/2},
	{src: 'photos/2.jpg', width: 400/2, height: 600/2},
	{src: 'photos/3.jpg', width: 480/2, height: 600/2},
	{src: 'photos/4.jpg', width: 600/2, height: 355/2},
	{src: 'photos/5.jpg', width: 600/2, height: 400/2},
	{src: 'photos/6.jpg', width: 600/2, height: 400/2},
	{src: 'photos/7.jpg', width: 600/2, height: 366/2}
]);

generateGrid('fixed-partition', {
	containerWidth: 1000,
	idealElementHeight: 150,
	spacing: 5,
	align: 'center'
}, [
	{src: 'photos/2.jpg', width: 400/2, height: 600/2},
	{src: 'photos/3.jpg', width: 480/2, height: 600/2}
]);

generateGrid('single', {
	containerWidth: 1000,
	maxHeight: 290,
	align: 'center'
}, [
	{src: 'photos/2.jpg', width: 400/2, height: 600/2}
]);
